blockchain/stakeversion_test.go

```diff
@@ -0,0 +1,805 @@
+func genesisBlockNode(params *chaincfg.Params) *blockNode {
+	// Create a new node from the genesis block.
+	genesisBlock := dcrutil.NewBlock(params.GenesisBlock)
+	header := &genesisBlock.MsgBlock().Header
+	node := newBlockNode(header, genesisBlock.Hash(), 0, []chainhash.Hash{},
+		[]chainhash.Hash{}, []uint32{})
+	node.inMainChain = true
+
+	return node
+}
+
+func TestCalcWantHeight(t *testing.T) {
+	// For example, if StakeVersionInterval = 11 and StakeValidationHeight = 13 the
+	// windows start at 13 + (11 * 2) 25 and are as follows: 24-34, 35-45, 46-56 ...
+	// If height comes in at 35 we use the 24-34 window, up to height 45.
+	// If height comes in at 46 we use the 35-45 window, up to height 56 etc.
+	tests := []struct {
+		name       string
+		skip       int64
+		interval   int64
+		multiplier int64
+		negative   int64
+	}{
+		{
+			name:       "13 11 10000",
+			skip:       13,
+			interval:   11,
+			multiplier: 10000,
+		},
+		{
+			name:       "27 33 10000",
+			skip:       27,
+			interval:   33,
+			multiplier: 10000,
+		},
+		{
+			name:       "mainnet params",
+			skip:       chaincfg.MainNetParams.StakeValidationHeight,
+			interval:   chaincfg.MainNetParams.StakeVersionInterval,
+			multiplier: 5000,
+		},
+		{
+			name:       "testnet params",
+			skip:       chaincfg.TestNetParams.StakeValidationHeight,
+			interval:   chaincfg.TestNetParams.StakeVersionInterval,
+			multiplier: 1000,
+		},
+		{
+			name:       "simnet params",
+			skip:       chaincfg.SimNetParams.StakeValidationHeight,
+			interval:   chaincfg.SimNetParams.StakeVersionInterval,
+			multiplier: 10000,
+		},
+		{
+			name:       "negative mainnet params",
+			skip:       chaincfg.MainNetParams.StakeValidationHeight,
+			interval:   chaincfg.MainNetParams.StakeVersionInterval,
+			multiplier: 1000,
+			negative:   1,
+		},
+	}
+
+	for _, test := range tests {
+		t.Logf("running: %v skip: %v interval: %v",
+			test.name, test.skip, test.interval)
+
+		start := int64(test.skip + test.interval*2)
+		expectedHeight := start - 1 // zero based
+		x := int64(0) + test.negative
+		for i := start; i < test.multiplier*test.interval; i++ {
+			if x%test.interval == 0 && i != start {
+				expectedHeight += test.interval
+			}
+			wantHeight := calcWantHeight(test.skip, test.interval, i)
+
+			if wantHeight != expectedHeight {
+				if test.negative == 0 {
+					t.Fatalf("%v: i %v x %v -> wantHeight %v expectedHeight %v\n",
+						test.name, i, x, wantHeight, expectedHeight)
+				}
+			}
+
+			x++
+		}
+	}
+}
+
+func newNode(blockVersion int32, height int64, currentNode *blockNode) *blockNode {
+	// Make up a header.
+	header := &wire.BlockHeader{
+		Version: blockVersion,
+		Height:  uint32(height),
+		Nonce:   0,
+	}
+	node := newBlockNode(header, &chainhash.Hash{}, 0,
+		[]chainhash.Hash{}, []chainhash.Hash{},
+		[]uint32{})
+	node.height = height
+	node.parent = currentNode
+
+	return node
+}
+
+func TestCalcStakeVersionCorners(t *testing.T) {
```

Comment:
This function works for now, but I personally would much rather see it table driven.  There is a ton of duplicated code in here that could easily be factored out.

I don't think converting it should be something required for release, but I'd like to see an issue opened for it so it's not forgotten.